### PR TITLE
Initial custom data type implementation

### LINKED ISSUE #8: phpType should guarantee the return type for all hydration types
especially returning as toArray().

There are a couple places we can do this. getBlah is one place, though I'm trying to find a sane place to do it on data retrieval instead. Depends on how consistent data comes back from the db.

### PR DESCRIPTION
**Pull request by bpicolo**
This fixes #8 by replacing phpType entirely

### CODE REVIEW

File: tests/Fixtures/bookstore/schema.xml [hunk 1]
```diff
@@ -400,4 +400,9 @@
         </foreign-key>
     </table>
 
+    <!-- Test custom phpType which is an object type -->
+    <table name="bookstore_custom_test">
+        <column name="id" primaryKey="true" autoIncrement="true" type="INTEGER"/>
+        <column name="test_datatype" type="VARCHAR" required="true" customType="Propel\Tests\Helpers\CustomDatabaseType"/>
```
Comment (LuqiPan) [inline tests/Fixtures/bookstore/schema.xml]: Nitpick: `test_datatype` made me think it's a type not a column. Something like `name="column_with_custom_database_type"`? (except there are too many characters)

Comment (bpicolo) [inline tests/Fixtures/bookstore/schema.xml]: Yeah, my naming suuucks. Will fix that

File: src/Propel/Runtime/CustomDatatypeInterface.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * in Propel, allowing you to have fine grained control over exactly what is
+ * going into your setters, how they get to the database, and how data flows from
+ * the database to column getters.
+ * Custom data types give you a simple strategy to extend propel to support
+ * data types such as Postgres json, hstore throughout your stack.
+ */
+interface CustomDataTypeInterface {
+    /**
+     * @param $instance - An instance of your custom type
+     * @return mixed
```
Comment (LuqiPan) [inline src/Propel/Runtime/CustomDatatypeInterface.php]: Can you add a description for the return value?

File: src/Propel/Runtime/CustomDatatypeInterface.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+     */
+    public static function __toDatabase($instance);
+
+    /**
+     * @param $data - The database representation of your objet
+     * @return mixed
```
Comment (LuqiPan) [inline src/Propel/Runtime/CustomDatatypeInterface.php]: Ditto to the return value here :smile:

**Comment by bpicolo** [comment]
merged this in alternative ways

**Pull request closed without merging**